File: typescript/design-as-you-code/CarWindshieldWiper.ts
import Car from "./Car";

const MIN_SPEED_LEVEL = 1;
const MAX_SPEED_LEVEL = 4;

export default class CarWindshieldWiper {
    private _isActive = false;
    private _speedContext: 'normal' | 'rain' = 'normal';
    private _speedLevel = 1;
    
    constructor(private car: Car) {}

    get isActive(): boolean {
        return this._isActive;
    }
    
    /** Expressed in ss (swipe by second) */
    get velocity(): number {
        if(!this._isActive){
            return 0;
        }

        const contextBaseVelocity = this._speedContext === 'normal' ? 0 : 4;

        return this._speedLevel + contextBaseVelocity;
    }

    set speedContext(value: 'normal' | 'rain') {
        this._speedContext = value;
    }

    activate() {
        if(!this.car.turnedOn) {
            return;
        }

        this._isActive = true;
    }

    deactivate() {
        this._isActive = false;
    }

    activateTemporarily(numberOfSwipes: number) {
        if(!this.car.turnedOn) {
            return;
        }

        const timeoutInSeconds = numberOfSwipes / this.velocity;

        this._isActive = true;

        setTimeout(this.deactivate, timeoutInSeconds * 1000);
    }

    increaseLevel() {
        if(!this._isActive){
            return;
        }

        const nextLevel = this._speedLevel + 1;

        if(nextLevel > MAX_SPEED_LEVEL){
            return;
        }

        this._speedLevel = nextLevel;
    }

    decreaseLevel() {
        if(!this._isActive){
            return;
        }

        const nextLevel = this._speedLevel - 1;

        if(nextLevel < MIN_SPEED_LEVEL){
            return;
        }

        this._speedLevel = nextLevel;
    }
}

File: python/src/domain/carWindshieldWiper.py
import asyncio

MIN_SPEED_LEVEL = 1
MAX_SPEED_LEVEL = 4

class CarWindshieldWiper:
    def __init__(self, car):
        self._car = car
        self._isActive = False
        self._speedContext = 'normal'
        """'normal' | 'rain'"""
        self._speedLevel = MIN_SPEED_LEVEL

    @property
    def isActive(self):
        return self._isActive
    
    @property
    def velocity(self):
        """Expressed in ss (swipe by second)"""

        if not self._isActive:
            return 0
        
        contextBaseVelocity = 0 if self._speedContext == 'normal' else 4

        return self._speedLevel + contextBaseVelocity

    @property
    def speedContext(self):
        return self._speedContext
    
    @speedContext.setter
    def speedContext(self, value):
        """'normal' | 'rain'"""
        self._speedContext = value

    def activate(self):
        if not self._car.turnedOn:
            print('entrou no pass')
            return

        self._isActive = True
    
    async def activateTemporarily(self, numberOfSwipes):
        self.activate()

        timeoutInSeconds = numberOfSwipes / self.velocity

        await asyncio.sleep(timeoutInSeconds)

        self.deactivate()

    def deactivate(self):
        self._isActive = False

    def increaseLevel(self):
        if not self._isActive:
            return
        
        newSpeedLevel = self._speedLevel + 1

        if newSpeedLevel > MAX_SPEED_LEVEL:
            return
        
        self._speedLevel = newSpeedLevel
    
    def decreaseLevel(self):
        if not self._isActive:
            return
        
        newSpeedLevel = self._speedLevel - 1

        if newSpeedLevel < MIN_SPEED_LEVEL:
            return
        
        self._speedLevel = newSpeedLevel